src/View/CastingService.php

```diff
@@ -0,0 +1,100 @@
+use SilverStripe\ORM\FieldType\DBInt;
+use SilverStripe\ORM\FieldType\DBText;
+
+class CastingService
+{
+    use Injectable;
+
+    /**
+     * Cast a value to the relevant object (usually a DBField instance) for use in the view layer.
+     *
+     * @param null|array|ModelData $source Where the data originates from. This is used both to check for casting helpers
+     * and to help set the value in cast DBField instances.
+     * @param bool $strict If true, an object will be returned even if $data is null.
+     */
+    public function cast(mixed $data, null|array|ModelData $source = null, string $fieldName = '', bool $strict = false): ?object
+    {
+        // null is null - we shouldn't cast it to an object, because that makes it harder
+        // for downstream checks to know there's "no value".
+        if (!$strict && $data === null) {
+            return null;
+        }
+
+        // Assume anything that's an object is intentionally using whatever class it's using
+        // and don't cast it.
+        if (is_object($data)) {
+            return $data;
+        }
+
+        $service = null;
+        if ($source instanceof ModelData) {
+            $service = $source->castingHelper($fieldName);
+        }
+
+        // Cast to object if there's an explicit casting for this field
+        // Explicit casts take precedence over array casting
+        if ($service) {
+            $castObject = Injector::inst()->create($service, $fieldName);
+            if (!ClassInfo::hasMethod($castObject, 'setValue')) {
+                throw new LogicException('Explicit casting service must have a setValue method.');
+            }
+            $castObject->setValue($data, $source);
+            return $castObject;
+        }
+
+        // Wrap arrays in ModelData so templates can handle them
+        if (is_array($data)) {
+            return array_is_list($data) ? ArrayList::create($data) : ArrayData::create($data);
+        }
+
+        // Fall back to default casting
+        $service = $this->defaultService($data, $source, $fieldName);
+        $castObject = Injector::inst()->create($service, $fieldName);
+        if (!ClassInfo::hasMethod($castObject, 'setValue')) {
+            throw new LogicException('Default service must have a setValue method.');
+        }
+        $castObject->setValue($data, $source);
+        return $castObject;
+    }
+
+    /**
+     * Get the default service to use if no explicit service is declared for this field on the source model.
+     */
+    private function defaultService(mixed $data, mixed $source = null, string $fieldName = ''): ?string
+    {
+        $default = null;
+        if ($source instanceof ModelData) {
+            $default = $source::config()->get('default_cast');
+            if ($default === null) {
+                $failover = $source->getFailover();
+                if ($failover) {
+                    $default = $this->defaultService($data, $failover, $fieldName);
```

Comment:
Could end up in an infinite loop here

Reply:
Failovers aren't allowed to reference themselves.
There are other places that also assume failovers aren't cyclical - including the code this replaces. This is probably the subject of another card to add cycle detection into `ModelData::setFailover()` if it's something you're worried about.

Reply:
OK leave as is then